pkg/upgrade/control_plane.go

```diff
@@ -334,18 +284,34 @@ func (u *ControlPlaneUpgrader) updateMachines(machines *clusterapiv1alpha1.Machi
 			return err
 		}
 
-		if err := u.applyAnnotation(&machine); err != nil {
+		if err := u.applyAnnotation(newMachine); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
 func (u *ControlPlaneUpgrader) applyAnnotation(m *clusterapiv1alpha1.Machine) error {
-	m.Annotations[UpgradeIDAnnotationKey] = u.upgradeID
-	// TODO(chuckha): use Patch instead of Update
-	_, err := u.managementClusterAPIClient.Machines(u.clusterNamespace).Update(m)
-	return errors.WithStack(err)
+	maxTries := 5
+	for i := 0; i < maxTries; i++ {
```

Comment:
I went with this to avoid importing controller runtime just for generating an object patch. I've not had much fun importing controller runtime in other (capi/capa) projects.

Reply:
We don't need controller-runtime for a patch. We can use client-go and json-patch.

Reply:
I haven't seen client-go export any feature that allows me to create a patch, can you point me to a function that client-go uses to generate patches or the specific json-patch library you're referring to. There seem to be a number of them.

Reply:
https://github.com/heptio/velero/blob/dd96aa76dbd1e3af36df6e60fab985d45b1d9d92/pkg/controller/download_request_controller.go#L226-L248

Reply:
oh. Do you mean generate a patch by hand that is a json-patch and then use client-go to call .Patch? That's a totally reasonable request

Reply:
Yes, you marshal before & after to json, then pass both jsons to `jsonpatch.CreateMergePatch`, then use client-go's Patch() call.

Reply:
yep, that non-k8s library was the missing piece.

Reply:
Updated!